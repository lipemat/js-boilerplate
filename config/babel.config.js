const {getBrowsersList} = require( '../helpers/config' );

const presetEnv = {
	bugfixes: true,
	corejs: {
		version: '3.31',
	},
	shippedProposals: true,
	targets: {
		browsers: getBrowsersList(),
	},
	useBuiltIns: 'usage',
};


module.exports = {
	cacheDirectory: true,
	presets: [
		[ '@babel/preset-env', presetEnv ],
		[ '@babel/preset-react', {
			development: 'production' !== process.env.NODE_ENV,
			runtime: 'automatic',
		} ],
		'@babel/preset-typescript',
	],
	plugins: [
		'@babel/plugin-syntax-dynamic-import',
	],
};
