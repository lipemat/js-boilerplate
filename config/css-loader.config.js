const {getLocalIdent, usingShortCssClasses} = require( '../helpers/css-classnames' );

/**
 * Options for the Webpack `css-loader`.
 *
 * Extracted to its on file to allow easy overrides.
 */

/**
 * Default to :global for classes in "global" or "pcss" directories.
 *
 * @param {string} resourcePath
 * @return {string}
 */
const mode = resourcePath => {
	if ( /globals?\//i.test( resourcePath.replace( /\\/g, '/' ) ) ) {
		return 'global';
	}
	if ( /pcss?\//i.test( resourcePath.replace( /\\/g, '/' ) ) ) {
		return 'global';
	}
	return 'local';
};

let cssLoader = {
	importLoaders: 1,
	modules: {
		exportLocalsConvention: 'camelCase',
		localIdentName: 'â“‚[name]__[local]__[contenthash:base64:2]',
		mode,
	},
	sourceMap: true,
	url: false,
};

if ( 'production' === process.env.NODE_ENV ) {
	cssLoader = {
		importLoaders: 1,
		modules: {
			exportLocalsConvention: 'camelCase',
			// Use short CSS Classes if enabled.
			...usingShortCssClasses() ? {getLocalIdent} : {},
			// Hash used when short CSS classes are not enabled.
			localIdentName: '[contenthash:base64:5]',
			mode,
		},
		url: false,
	};
}

module.exports = cssLoader;
