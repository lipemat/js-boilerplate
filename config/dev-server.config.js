const fs = require( 'fs' );
const packageConfig = require( '../helpers/package-config' );

const url = new URL( packageConfig.url );

let server = 'https:' === url.protocol ? 'https' : 'http';
// Load local certificates for https during development.
if ( 'object' === typeof ( packageConfig.certificates ) ) {
	server = {
		type: 'https',
		options: {
			cert: fs.readFileSync( packageConfig.certificates.cert ),
			key: fs.readFileSync( packageConfig.certificates.key ),
		},
	};
}

module.exports = {
	allowedHosts: 'all',
	client: {
		logging: 'warn',
		overlay: {
			errors: true,
			warnings: false,
		},
	},
	headers: {
		'Access-Control-Allow-Origin': '*',
		'Access-Control-Allow-Methods': '*',
		'Access-Control-Allow-Headers': '*',
	},
	host: '0.0.0.0', // Allow connections from any IP.
	hot: true,
	port: 3000,
	server,
	static: false,
};
