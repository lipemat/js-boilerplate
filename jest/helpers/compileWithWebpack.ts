import {basename, extname, join, resolve} from 'path';
import webpack, {type Compiler} from 'webpack';
import 'setimmediate';
import {createFsFromVolume, IFs, Volume} from 'memfs';

const {getConfig} = require( '../../helpers/config' );

/**
 * Compile a file using webpack.
 *
 *
 */


export type Fixture = {
	input: string;
	output: string;
	basename: string;
	description: string;
}

/**
 * Mimic webpack compiler.run() method with promises support.
 * This is a workaround for webpack does not support promises.
 *
 */
function compile( compiler: Compiler, fixture: Fixture ): Promise<string> {
	return new Promise( ( resolve, reject ) => {
		compiler.run( ( error, stats ) => {
			if ( error || 'undefined' === typeof stats ) {
				reject( error );
				return;
			}
			if ( stats.hasErrors() ) {
				reject( stats.compilation.errors );
			}

			try {
				const usedFs = compiler.outputFileSystem as IFs;
				const outputPath = stats.compilation.outputOptions.path ?? '';
				const data = usedFs.readFileSync( join( outputPath, basename( fixture.output ) ) ).toString();
				resolve( data.trim() );
			} catch ( e ) {
				if ( stats.compilation.errors.length > 0 ) {
					reject( stats.compilation.errors );
				}
				reject( e );
			}
		} );
		setTimeout( () => {
			reject( new Error( 'Timed out' ) );
		}, 2_000 );
	} );
}

/**
 * Compile a fixture with webpack.
 *
 * @notice We can't reset the modules of MiniCssExtractPlugin conflicts with
 *         itself. Instead, we isolate other configurations to allow them to
 *         load fresh each time.
 *
 */
export default function compileWithWebpack( fixture: Fixture, config = {} ): Promise<string> {
	const fullConfig = {...require( '../../config/webpack.dist.js' ), ...config};

	// Isolate the css-loader and postcss config, so it is loaded fresh each time.
	// Allow differenicate between production and development.
	jest.isolateModules( () => {
		fullConfig.module.rules[ 2 ].use[ 1 ].options = getConfig( 'css-loader.config' );
		fullConfig.module.rules[ 2 ].use[ 2 ].options.postcssOptions = getConfig( 'postcss.config' );
	} );

	// Point a single entry to the fixture file.
	const entry = basename( fixture.input, extname( fixture.input ) );
	fullConfig.entry = {
		[ entry ]: resolve( __dirname, '../../', fixture.input ),
	};
	fullConfig.context = resolve( __dirname, '../fixtures' );

	// The manifest plugin does not work in this context.
	delete fullConfig.plugins[ 6 ];

	// Create a compiler with the fixture config.
	const compiler = webpack( fullConfig );

	// Use a memory cache for the compiler file system.
	compiler.outputFileSystem = createFsFromVolume( new Volume() );

	return compile( compiler, fixture );
}
