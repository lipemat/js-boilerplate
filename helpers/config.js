const packageConfig = require( './package-config' );
const path = require( 'path' );
const fs = require( 'fs' );
const once = require( 'lodash/once' );
const browserslist = require( 'browserslist' );

const extensions = [
	...Object.keys( packageConfig.dependencies ?? {} ).filter( name => name.includes( 'js-boilerplate-' ) ),
	...Object.keys( packageConfig.devDependencies ?? {} ).filter( name => name.includes( 'js-boilerplate-' ) ),
];

/**
 * Check to see if a local config file exists.
 *
 * @param {string}  fileName
 * @param {boolean} inWorkingDirectory - Look in working directory instead of their /config directory
 *
 * @return {boolean}
 */
function hasLocalOverride( fileName, inWorkingDirectory = false ) {
	let hasLocal = false;
	try {
		if ( inWorkingDirectory ) {
			require( path.resolve( packageConfig.workingDirectory, fileName ) );
			hasLocal = true;
		} else {
			require( path.resolve( packageConfig.packageDirectory + '/config', fileName ) );
			hasLocal = true;
		}
	} catch ( e ) {
	}

	return hasLocal;
}

/**
 * Get a config from our /config directory merged with any
 * matching configuration from the project directory.
 *
 * For instance if we have a file named config/babel.config.js in our project
 * we will merge the contents with our config/babel.config.js in favor of whatever
 * is specified with the project's file.
 *
 * If the `module.exports` are a function, the existing configuration will be passed
 * as the only argument. Otherwise, standard `module.exports` are also supported.
 *
 * @example ```ts
 * // standard
 * module.export = {
 *     externals: {extra: 'Extra'}
 * }
 * // function
 * module.exports = function( config ) {
 *     return {
 *         externals: {...config.externals, extra: 'Extra'}
 *     }
 * }
 * ```
 *
 * @param {string} fileName
 *
 * @return {Object}
 */
function getConfig( fileName ) {
	let config = require( '../config/' + fileName );
	config = {...config, ...getExtensionsConfig( fileName, config )};
	try {
		const localConfig = require( path.resolve( packageConfig.packageDirectory + '/config', fileName ) );
		if ( 'function' === typeof localConfig ) {
			config = {...config, ...localConfig( config )};
		} else {
			config = {...config, ...localConfig};
		}
	} catch ( e ) {
	}
	return config;
}

/**
 * Get a config from any existing extension's /config directories
 * merged into one.
 *
 * @param {string} fileName
 * @param {Object} defaultConfig - Default config from this package.
 *                               Used for passing to an extension callback.
 *
 * @see getConfig
 *
 * @return {Object}
 */
function getExtensionsConfig( fileName, defaultConfig ) {
	let config = {};
	extensions.forEach( extension => {
		try {
			const extensionConfig = require( extension + '/config/' + fileName );
			if ( 'function' === typeof extensionConfig ) {
				config = {...config, ...extensionConfig( {...defaultConfig, ...config} )};
			} else {
				config = {...config, ...extensionConfig};
			}
		} catch ( e ) {
		}
	} );

	return config;
}

/**
 * Get the path to the "tsconfig.json" file if it exists.
 *
 * 1. The working directory.
 * 2. The package directory.
 *
 * The package directory takes priority over the working directory.
 *
 *
 * @return {string}
 */
function getTsConfigFile() {
	const possibles = [
		// Backward compatible for before @lipemat/eslint-config version 3.
		path.resolve( packageConfig.workingDirectory + '/tsconfig.json' ),
		path.resolve( packageConfig.packageDirectory + '/tsconfig.json' ),
	];
	let tsConfig = '';
	possibles.forEach( filePath => {
		if ( fs.existsSync( filePath ) ) {
			tsConfig = filePath;
		}
	} );
	return tsConfig;
}

/**
 * If browserslist is not specified, we fall back to WordPress defaults
 * except for > 1% we don't support by default.
 *
 * Return false if a browserslist is specified in the current project.
 *
 * @link https://github.com/browserslist/browserslist#config-file
 *
 * @return {boolean | string[]}
 */
const getDefaultBrowsersList = once( () => {
	// JEST requires the default browsers list for JSDOM to work.
	if ( typeof global.__TEST__ !== 'undefined' && global.__TEST__ ) {
		return false;
	}

	if ( browserslist( browserslist.defaults ) === browserslist() ) {
		return require( '@wordpress/browserslist-config' ).map( range => {
			// Swap out "> 1%" for "> 2%".
			return '> 1%' === range ? '> 2%' : range;
		} );
	}
	return false;
} );


module.exports = {
	getConfig,
	getDefaultBrowsersList,
	getExtensionsConfig,
	hasLocalOverride,
	getTsConfigFile,
};
