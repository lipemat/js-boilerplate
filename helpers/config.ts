import {existsSync} from 'fs';
import {resolve} from 'path';
import webpack from 'webpack';
import type {BabelConfig} from '../config/babel.config';
import type {JestConfig} from '../config/jest.config';
import {getPackageConfig} from './package-config';
import type {EntriesConfig} from '../config/entries.config';
import type {Config as PostCSSConfig} from 'postcss-load-config';
import type {CssLoaderConfig} from '../config/css-loader.config';

// Must be required to avoid issues with browserslist.
const browserslist = require( 'browserslist' );


type Configs = {
	'babel.config': BabelConfig;
	'css-loader.config': CssLoaderConfig;
	'entries.config': EntriesConfig;
	'jest.config': JestConfig;
	'postcss.config': PostCSSConfig;
	'webpack.dist': webpack.Configuration;
};

const {dependencies, devDependencies, workingDirectory, packageDirectory} = getPackageConfig();

const extensions = [
	...Object.keys( dependencies ?? {} ).filter( name => name.includes( 'js-boilerplate-' ) ),
	...Object.keys( devDependencies ?? {} ).filter( name => name.includes( 'js-boilerplate-' ) ),
];


/**
 * Check to see if a local config file exists.
 *
 * @param {string}  fileName
 * @param {boolean} inWorkingDirectory - Look in working directory instead of their /config directory
 *
 * @return {boolean}
 */
export function hasLocalOverride( fileName: string, inWorkingDirectory: boolean = false ): boolean {
	let hasLocal = false;
	try {
		if ( inWorkingDirectory ) {
			require( resolve( workingDirectory, fileName ) );
			hasLocal = true;
		} else {
			require( resolve( packageDirectory + '/config', fileName ) );
			hasLocal = true;
		}
	} catch ( e ) {
		if ( 'MODULE_NOT_FOUND' !== e.code ) {
			console.error( e );
		}
	}

	return hasLocal;
}


/**
 * Get a config from our /config directory merged with any
 * matching configuration from the project directory.
 *
 * For instance if we have a file named config/babel.config.js in our project
 * we will merge the contents with our config/babel.config.js in favor of whatever
 * is specified with the project's file.
 *
 * If the `module.exports` are a function, the existing configuration will be passed
 * as the only argument. Otherwise, standard `module.exports` are also supported.
 *
 * @example ```ts
 * // standard
 * module.export = {
 *     externals: {extra: 'Extra'}
 * }
 * // function
 * module.exports = function( config ) {
 *     return {
 *         externals: {...config.externals, extra: 'Extra'}
 *     }
 * }
 * ```
 *
 * @param {string} fileName
 *
 * @return {Object}
 */
export function getConfig<T extends keyof Configs>( fileName: T ): Configs[T] {
	let mergedConfig = require( '../config/' + fileName ) as Configs[T];
	mergedConfig = {...mergedConfig, ...getExtensionsConfig<Configs[T]>( fileName, mergedConfig )};
	try {
		const localConfig = require( resolve( packageDirectory + '/config', fileName ) );
		if ( 'function' === typeof localConfig ) {
			mergedConfig = {...mergedConfig, ...localConfig( mergedConfig )};
		} else {
			mergedConfig = {...mergedConfig, ...localConfig};
		}
	} catch ( e ) {
		if ( 'MODULE_NOT_FOUND' !== e.code ) {
			console.error( e );
		}
	}
	return mergedConfig;
}


/**
 * Get a config from any existing extension's /config directories
 * merged into one.
 *
 * @param {string} fileName
 * @param {Object} defaultConfig - Default config from this package.
 *                               Used for passing to an extension callback.
 *
 * @see getConfig
 *
 * @return {Object}
 */
export function getExtensionsConfig<T extends object>( fileName: string, defaultConfig: T ): T {
	let mergedConfig: T = {} as T;
	extensions.forEach( extension => {
		try {
			const extensionConfig = require( extension + '/config/' + fileName );
			if ( 'function' === typeof extensionConfig ) {
				mergedConfig = {...mergedConfig, ...extensionConfig( {...defaultConfig, ...mergedConfig} )};
			} else {
				mergedConfig = {...mergedConfig, ...extensionConfig};
			}
		} catch ( e ) {
			if ( 'MODULE_NOT_FOUND' !== e.code ) {
				console.error( e );
			}
		}
	} );

	return mergedConfig;
}


/**
 * Get the path to the "tsconfig.json" file if it exists.
 *
 * 1. The working directory.
 * 2. The package directory.
 *
 * The package directory takes priority over the working directory.
 *
 *
 * @return {string}
 */
export function getTsConfigFile(): string {
	const possibles = [
		// Backward compatible for before @lipemat/eslint-config version 3.
		resolve( workingDirectory + '/tsconfig.json' ),
		resolve( packageDirectory + '/tsconfig.json' ),
	].filter( existsSync );

	let tsConfig = '';
	possibles.forEach( filePath => {
		tsConfig = filePath;
	} );
	return tsConfig;
}


/**
 * Get the browserslist from the current project.
 *
 * - If specified using standard browserslist config, we will use that.
 *
 *  @link https://github.com/browserslist/browserslist#config-file
 */
export function getBrowsersList(): readonly string[] {
	const projectBrowsersList = browserslist();
	if ( browserslist( browserslist.defaults ) === projectBrowsersList ) {
		const wp = [ ...require( '@wordpress/browserslist-config' ) ];
		return adjustBrowserslist( wp );
	}
	return projectBrowsersList;
}


/**
 * If browserslist is not specified, we fall back to WordPress defaults.
 *
 * - Return the default browserslist if the current project does not specify one.
 * - Return false if a browserslist is specified.
 *
 * Used in cases where we can fall back to standard browserslist config if the project
 * has not specified one.
 *
 * @deprecated Use getBrowsersList instead.
 *
 * @link https://github.com/browserslist/browserslist#config-file
 *
 * @return {boolean | string[]}
 */
export const getDefaultBrowsersList = (): false | string[] => {
	if ( browserslist( browserslist.defaults ) === browserslist() ) {
		const wp = [ ...require( '@wordpress/browserslist-config' ) ];
		return adjustBrowserslist( wp );
	}
	return false;
};


/**
 * Adjust the browserslist to include our defaults.
 *
 * @todo Remove `not op_mini all` after 3/8/2024 if it does not creep back in to the defaults.
 */
export function adjustBrowserslist( browserRules: string[] ): string[] {
	browserRules.push( 'not op_mini all' );
	return browserRules;
}
