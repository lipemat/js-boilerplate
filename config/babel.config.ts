import {getBrowsersList} from '../helpers/config';
import type {TransformOptions} from '@babel/core';
import type {Options} from '@babel/preset-env';

/**
 * @link https://webpack.js.org/loaders/babel-loader/#options
 */
export type BabelLoader = {
	cacheDirectory?: boolean | string;
	cacheIdentifier?: string;
	cacheCompression?: boolean;
	customize?: string;
	metadataSubscribers?: string[];
}

/**
 * Use Babel's preset-env to add support for target browsers.
 *
 * @note Set the `debug` option to `true` to debug the included polyfills and plugins.
 *
 * @see https://babeljs.io/docs/en/babel-preset-env
 */
const presetEnv: Options = {
	bugfixes: true,
	corejs: {
		// Use the core-js version currently installed in the project.
		version: require( 'core-js/package.json' ).version,
		proposals: false,
	},
	// Enable the `debug` option to debug the included polyfills and plugins.
	debug: false,
	// Ignore any external browserslist in favor of `getBrowsersList()`.
	ignoreBrowserslistConfig: true,
	shippedProposals: false,
	targets: {
		browsers: getBrowsersList(),
	},
	useBuiltIns: 'usage',
};


const babelConfig: TransformOptions & BabelLoader = {
	cacheDirectory: true,
	presets: [
		[ '@babel/preset-env', presetEnv ],
		[ '@babel/preset-react', {
			development: 'production' !== process.env.NODE_ENV,
			runtime: 'automatic',
		} ],
		'@babel/preset-typescript',
	],
	plugins: [
		'@babel/plugin-syntax-dynamic-import',
	],
};

module.exports = babelConfig;
