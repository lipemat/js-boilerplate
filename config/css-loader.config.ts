import {getLocalIdent, usingShortCssClasses} from '../helpers/css-classnames';
import type {Config, Mode} from 'css-loader';
import type {AtLeast} from '../types/utility';

export type CssLoaderConfig = AtLeast<Config, 'importLoaders' | 'modules' | 'sourceMap' | 'url'>

/**
 * Options for the Webpack `css-loader`.
 *
 * Extracted to its on file to allow easy overrides.
 */

/**
 * Default to :global for classes in "global" or "pcss" directories.
 *
 * @param {string} resourcePath
 * @return {string}
 */
const mode = ( resourcePath: string ): Mode => {
	if ( /globals?\//i.test( resourcePath.replace( /\\/g, '/' ) ) ) {
		return 'global';
	}
	if ( /pcss?\//i.test( resourcePath.replace( /\\/g, '/' ) ) ) {
		return 'global';
	}
	return 'local';
};

const cssLoader: CssLoaderConfig = {
	importLoaders: 1,
	modules: {
		exportLocalsConvention: 'camelCase',
		localIdentName: 'â“‚[name]__[local]__[contenthash:base64:2]',
		mode,
	},
	sourceMap: true,
	url: false,
};

if ( 'production' === process.env.NODE_ENV ) {
	cssLoader.modules = {
		exportLocalsConvention: 'camelCase',
		// Use short CSS Classes if enabled.
		...usingShortCssClasses() ? {getLocalIdent} : {},
		// Hash used when short CSS classes are not enabled.
		localIdentName: '[contenthash:base64:5]',
		mode,
	}
	cssLoader.sourceMap = false;
}

module.exports = cssLoader;
