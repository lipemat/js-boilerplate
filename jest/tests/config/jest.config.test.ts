import {getConfig} from '../../../helpers/config';

describe( 'jest.config.test.ts', () => {
	it( 'should match the snapshot', () => {
		const config = require( '../../../config/jest.config' );
		expect( config ).toMatchSnapshot();
		expect( config ).toStrictEqual( getConfig( 'jest.config' ) );
	} );

	test( 'Build files', () => {
		const TS = require( '../../../config/jest.config.ts' );
		const JS = require( '../../../config/jest.config.js' );
		expect( TS ).toStrictEqual( JS );
	} );
} );
