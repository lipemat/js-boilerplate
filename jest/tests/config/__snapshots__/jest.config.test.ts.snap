// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jest.config.test.ts should match the snapshot 1`] = `
{
  "globals": {
    "__TEST__": true,
  },
  "moduleNameMapper": {
    "\\.(pcss|less|css)$": "identity-obj-proxy",
    "is-plain-obj": "identity-obj-proxy",
    "uuid": "identity-obj-proxy",
  },
  "roots": [
    "./tests",
  ],
  "setupFilesAfterEnv": [],
  "testEnvironment": "jsdom",
  "testEnvironmentOptions": {
    "url": "http://localhost",
  },
  "transform": {
    "^.+\\.[tj]sx?$": [
      "babel-jest",
      {
        "plugins": [
          "@babel/plugin-syntax-dynamic-import",
        ],
        "presets": [
          [
            "@babel/preset-env",
            {
              "bugfixes": true,
              "corejs": {
                "proposals": false,
                "version": "3.37.1",
              },
              "debug": false,
              "ignoreBrowserslistConfig": true,
              "shippedProposals": false,
              "targets": {
                "browsers": [
                  "> 1%",
                  "last 1 Android versions",
                  "last 1 ChromeAndroid versions",
                  "last 2 Chrome versions",
                  "last 2 Firefox versions",
                  "last 2 Safari versions",
                  "last 2 iOS versions",
                  "last 2 Edge versions",
                  "last 2 Opera versions",
                  "not op_mini all",
                ],
              },
              "useBuiltIns": "usage",
            },
          ],
          [
            "@babel/preset-react",
            {
              "development": true,
              "runtime": "automatic",
            },
          ],
          "@babel/preset-typescript",
        ],
      },
    ],
  },
  "transformIgnorePatterns": [
    "node_modules/(?!@lipemat)",
  ],
}
`;
